inspirehep/modules/records/api.py

```diff
@@ -95,6 +152,243 @@ def validate(self):
         """Validate the record, also ensuring format compliance."""
         validate(self)
 
+    def add_document_or_figure(
+        self,
+        metadata,
+        stream=None,
+        is_document=True,
+        file_name=None,
+        key=None,
+    ):
+        if not key and not file_name:
+            raise TypeError(
+                'No file_name and no key passed, at least one of them is '
+                'needed.'
+            )
+
+        if not key:
+            key = self._get_unique_files_key(base_file_name=file_name)
+
+        if stream is not None:
+            self.files[key] = stream
+
+        builder = LiteratureBuilder(record=self.dumps())
+        metadata['key'] = key
+        metadata['url'] = '/api/files/{bucket}/{key}'.format(
+            bucket=self.files[key].bucket_id,
+            key=key,
+        )
+        if is_document:
+            builder.add_document(**metadata)
+        else:
+            builder.add_figure(**metadata)
+
+        super(InspireRecord, self).update(builder.record)
+        return metadata
+
+    def _resolve_doc_or_fig_url(
+        self,
+        doc_or_fig_obj,
+        src_record=None,
+        only_new=False,
+    ):
+        """Resolves the given document url according to the current record.
+
+        It will fill it up with the local url in case it's found in the given
+        src_record.
+        It will leave any document that does not need to be downloaded with the
+        url intact, pointing to the local files api endpoint.
+        """
+        doc_or_fig_obj = copy.deepcopy(doc_or_fig_obj)
+        key = doc_or_fig_obj['key']
+
+        if not doc_or_fig_obj['url'].startswith('/api/files/'):
```

Comment:
So this method only does something for `invenio-files` URLs. What is the return value and why is so much code needed when the only line that seems to 'do something' is [this one](https://github.com/inspirehep/inspire-next/pull/2899/files#diff-3cf20a9f11cd40bad5e31856ef439299R239)?

Reply:
It tries to resolve the `/api/files` url's, and raises the correct exceptions if unable, properly hinting on what is wrong with the data passed, that's why there are so many ifs, otherwise it will raise only one generic exception with no hint.
Taking into account too if this is just an update or not (`update_only` param).

Then, in the case that you passed a `src_record`, and the `key` is not in `self.files` but it's on `src_record`, it uses the local path to that file in the returned document metadata, so later, we can download it safely.

The return object is a dict with:
* In the case it has to be downloaded: the metadata of the document that we can actually download (the `url` is set to something that you can use `open_url_or_path` on). 
* In the case it does not have to be downloaded: the metadat of the document with a local api `url` (/api/files/...).

I'll clarify the docs.

The alternative code with no hints would be:
```python
if only_new and src_record and key in src_record and key not in self.files:
    doc_or_fig_obj['url'] = src_record.files[key].file.uri
    return doc_or_fig_obj

if not src_record and only_new and key in self.files:
    return doc_or_fig_obj

raise Exception("There's something wrong with the files.")
```
but the conditions being that 'complex' I preferred raising a sensible exception.

Hmm... I can split the two things, something like:
```python
if only_new and src_record and key in src_record and key not in self.files:
    doc_or_fig_obj['url'] = src_record.files[key].file.uri
    return doc_or_fig_obj

if not src_record and only_new and key in self.files:
    return doc_or_fig_obj

### then after this, try to give a nice exception
if not src_record and not only_new:
    raise Exception('you did not pass a record to get files from...')

if not src_record  and only_update and key not in self.files:
    raise Exception('Bad document ...')    
```

Hmm... refactoring!